### ISSUE TITLE
Update using arrayFilters throws error

### ISSUE THREAD

**Issue by Cherviakov** [Automattic/mongoose#7431]
**Do you want to request a *feature* or report a *bug*?**
Bug
**What is the current behavior?**
Suppose we have schema with field 
```javascript
itemsInfo: {
  allUsers: { all: Number },
  individual: [{
    userId: String,
    all: Number,
  }]
}
```
We want to update just fields itemsInfo.individual.all, where userId belongs to certains array. Accoring do [mongodb docs](https://docs.mongodb.com/manual/reference/operator/update/positional-filtered/) it is possible to do with specifying arrayFilters in update. Current latest version of mongoose (mongoose 5.4.5) it throws error:
ReferenceError: context is not defined
    at castFilterPath (myProject/node_modules/mongoose/lib/helpers/query/castFilterPath.js:48:18)
    at castArrayFilters myProject/node_modules/mongoose/lib/helpers/update/castArrayFilters.js:57:26)
    at _castArrayFilters (myProject/node_modules/mongoose/lib/query.js:1739:5)
    at model.Query._updateThunk (myProject/node_modules/mongoose/lib/query.js:3445:3)
    at model.Query.<anonymous> (myProject/node_modules/mongoose/lib/query.js:3548:23)
    at model.Query._wrappedThunk [as _updateOne] (myProject/node_modules/mongoose/lib/helpers/query/wrapThunk.js:16:8)
    at process.nextTick (myProject/node_modules/kareem/index.js:369:33)
    at process.internalTickCallback (internal/process/next_tick.js:70:11)

**If the current behavior is a bug, please provide the steps to reproduce.**

```javascript
myModel.updateOne(
        { _id: container._id },
        {
          $inc: {
            'itemsInfo.allUsers.all': 1,
            'itemsInfo.individual.$[element].all': 1
          }
        },
        { arrayFilters: [{ 'element.userId': { $in: item.users } }] }
 );
```
**What is the expected behavior?**
Work not throwing error
**Please mention your node.js, mongoose and MongoDB version.**
nodejs: 11.1.0
mongoose: 5.4.5
mongodb: 4.0.5

**Comment by vkarpov15** [comment]
The below script repos this issue:

```javascript
const assert = require('assert');
const mongoose = require('mongoose');
mongoose.set('debug', true);

const GITHUB_ISSUE = `gh7431`;
const connectionString = `mongodb://localhost:27017/${ GITHUB_ISSUE }`;
const { Schema } = mongoose;

run().then(() => console.log('done')).catch(error => console.error(error.stack));

async function run() {
  await mongoose.connect(connectionString);
  await mongoose.connection.dropDatabase();

  const schema = new Schema({
    itemsInfo: {
      allUsers: { all: Number },
      individual: [{
        userId: String,
        all: Number,
      }]
    }
  });

  const Model = mongoose.model('gh7431', schema);

  await Model.updateOne(
        {},
        {
          $inc: {
            'itemsInfo.allUsers.all': 1,
            'itemsInfo.individual.$[element].all': 1
          }
        },
        { arrayFilters: [{ 'element.userId': { $in: ['1', '2', '3'] } }] }
  );
}
```

**Comment by vkarpov15** [comment]
Fixed in master, will be released in 5.4.8. We got bit by [this issue](https://github.com/eslint/eslint/issues/11072): enabling mocha mode in eslint means `context` is a global variable, so our tests and linter were passing. I also fixed our eslint config to catch this going forward.

**Comment by Cherviakov** [comment]
Much obliged!
